### ISSUE TITLE
Server-rendered HTML minification

### ISSUE THREAD

**Issue by tigt** [marko-js/marko#1432]
## Description

Marko’s compilation to server-side templates could perform provably-safe HTML minification during compilation. (Runtime minification is probably too expensive to be worth it.)

## Why

Perf matters! Running Marko’s current home page through [kangax’s minifier](https://kangax.github.io/html-minifier/) saves 2,551 bytes. Not too bad after compression, but it’s still overhead, and fewer characters speed up parsing.

<details>
I used the relative URLs option, kept comments because I suspected Marko uses them for rehydration, and turned off the Minify CSS/JS options.
</details>

Existing HTML minifiers _can_ be plugged into a Marko app’s render pipeline, but they require buffering the stream, defeating any performance improvements.

## Possible Implementation & Open Questions

In development mode, probably don’t bother running this compilation step. But in production…

+ Normalize output for better compression
  - Normalize known tag/attribute name casing (might do this already?)
  - Sort attributes (probably alphabetically, but frequency would be even better)
  - Sort properties in `style`
  - Sort tokens in `class`, `rel`, and other TokenLists/similar
+ Collapse whitespace (Marko largely already does this, but it’s worth reexamining to ensure its behavior is as unsurprising and efficient as it could be)
+ Collapse all whitespace characters to a single space except when `<pre>`, `marko-preserve-whitespace`
+ Omit the closing slash on known void elements
+ Optional tags
  - Omit optional opening tags without attributes like `<head>` or `<body>`
  - Omit optional closing tags where safe — `</body></html>`, consecutive `<p>`s, `</option>` since `</select>` autocloses it
+ Omit attributes with default values :`method="get"`, `<input type="text">`, `<script type="text/javascript">`, etc. — this would also make rendering client-side slightly more efficient
+ Attribute values
  - Truthy booleans should omit `=""`
  - Literals not containing whitespace should omit quotes
  - Literals containing `"` or `'`' should use the other delimiter for the value
  - If the attribute value is known not to stringify with forbidden characters (`Number`, for example), it should omit quotes
  - Everything else should fall back to wrapping double quotes
+ Reencode character references
  - Non-HTML syntax characters like `&rarr;` should be dereferenced
  - Unnecessary encoding should be dereferenced
    * Unambiguous ampersands
    * Quotes and `>` outside opening tag contexts
    * `=` outside attribute values
  - Some character references could be reencoded to be shorter:
    * `&#60;` &rarr; `&lt;` (pretty sure only `<` and `>` qualify)
    * `&#x27;` &rarr; `&#39;`, since hex numbers are only more compact for characters we’re already decoding
    * `&quot;` &rarr; `&#34;`
+ Turn same-origin attribute values with URLs (`href`, `src`, `srcset`) into relative URLs
+ `data-marko` and `data-marko-key` attributes could be more compact
   - `data-marko-key="@input s0-0-8-14-4"` could become `data-marko-key=input@Saaimd` (intentionally designed to allow omitting quotes, numbers encoded as an offset in ASCII). This would also make `<!--M^-->` comments and keys inside `data-marko` more compact, and a good string encoding algorithm could be faster than the current hyphen-separated numbers.
   - `data-marko` is currently JSON, which is probably the most lightweight way of decoding an object in the browser. That said…
      ```
      data-marko='{"onclick":"focus s0-0-8-4-4 false","onsubmit":"search s0-0-8-4-4 false"}'
     ```
     This object looks like it has restrictions that make it less flexible than a JS object. So we could try something like…

     ```
     data-marko=click:focus-Saaidd-0,submit:search-Saaidd-0
     ```
     Removes `on` prefix, uses shorter `key` encoding, `false` as `0`, and strips JSON delimiters. I suspect the `false` parameter could be omitted when it’s not true as well.

## Is this something you're interested in working on?

Yes

**Comment by tigt** [comment]
Thinking more about it, this is my current idea for `data-marko-key`:

Numbers are a single UTF-8 character, using `String.fromCodePoint(num + 97)` to ensure no ASCII whitespace (since that’s the only thing forbidden in HTML’s `id` attribute) and the iffy-inside-attributes backtick. This lets each character be its own boundary, so no need for `-`.

That should work for lengths as large as the Basic Multilingual Plane (since JS has trouble with outside that), which frankly would be a bad idea for other reasons to have Marko render _that much_.

If any separators are needed (like `[]` in the current keys), that’s what [the ASCII field separators](https://en.wikipedia.org/wiki/C0_and_C1_control_codes#Field_separators) were literally made for:

Name | Code | Index
------|-------|-------
File Separator | `\x1c` | 28
Group Separator | `\x1d` | 29
Record Separator | `\x1e` | 30
Unit Separator | `\x1f` | 31

Since their character indices are smaller than 97, there won’t be collisions. If that’s too unreadable, then all the ASCII punctuation Marko already uses is below 97 anyway.

**Comment by tigt** [comment]
Some updated thoughts on my proposals above:

> * Normalize output for better compression
>     * Normalize known tag/attribute name casing (might do this already?)
>     * Sort attributes (probably alphabetically, but frequency would be even better)
>     * Sort properties in `style`
>     * Sort tokens in `class`, `rel`, and other TokenLists/similar

While `DOMTokenList`s produced from attributes like `class` and `rel` _are_ order-agnostic, author CSS and JS may rely on specific order (mostly attribute selectors like `[class^=…]`). The compression gains from ordering these tokens are unlikely to be significant enough for that risk.

> * Omit optional closing tags where safe — `</body></html>`, consecutive `<p>`s, `</option>` since `</select>` autocloses it
> 
> * Omit attributes with default values: `method="get"`, `<input type="text">`, `<script type="text/javascript">`, etc. — this would also make rendering client-side slightly more efficient

After some more investigation, these two seem like they have the best reduction:effort ratio. Lotta `</option>`s in most `<select>`s, and HTML minification doubling as less JS is always nice.

> * Turn same-origin attribute values with URLs (`href`, `src`, `srcset`) into relative URLs

This one seems like the most difficult, but it also can pay off the most. I’m undecided.

> * `data-marko` and `data-marko-key` attributes could be more compact
>   * `data-marko-key="@input s0-0-8-14-4"` could become `data-marko-key=input@Saaimd` (intentionally designed to allow omitting quotes, numbers encoded as an offset in ASCII). This would also make `<!--M^-->` comments and keys inside `data-marko` more compact, and a good string encoding algorithm could be faster than the current hyphen-separated numbers.
>   * `data-marko` is currently JSON, which is probably the most lightweight way of decoding an object in the browser. That said…
>       ```
>       data-marko='{"onclick":"focus s0-0-8-4-4 false","onsubmit":"search s0-0-8-4-4 false"}'
>       ```
>         
>       This object looks like it has restrictions that make it less flexible than a JS object. So we could try something like…
>       ```
>       data-marko=click:focus-Saaidd-0,submit:search-Saaidd-0
>       ```

Spitballing this approach with the eBay homepage makes it seem quite appealing, but it’s also the one part where behavior isn’t from the HTML Standard. What do the Marko maintainers think?

**Comment by alexnewmannn** [comment]
@tigt Hey, just been looking through this issue and a few MR's around the unquoted attributes. We upgraded to latest and tracked our issues to the unquoted attributes. I don't feel like this requires an issue, but let me know if you prefer it in that format. Is there a way to remove this behaviour or toggle it on/off? We use a HTML to PDF converter, where our HTML is generated by Marko, and our PDF converter that we are using doesn't parse anything with unquoted attributes due to the type of spec it adheres to so right now we are stuck at 4.20.2 for the foreseeable.

**Comment by tigt** [comment]
@alexnewmannn I don’t believe as it stands this feature has a config toggle. (I didn’t implement it, just yammered about it in this issue.)

For a fix that would solve your problem right now, you could round-trip the HTML Marko produces into and then out of a HTML parsing library, then pass _that_ to your PDF converter. To do that with the `parse5` package would look something like:

```js
import { parse, serialize } from 'parse5'

const minifiedHtml = markoTemplate.render(…) // or however you do it

const normalizedHtml = serialize(parse(minifiedHtml))
```

**Comment by tigt** [comment]
Removed the stuff about smaller internal ID encodings since any element index between 30 and ~50 would encode as forbidden Latin-1 Supplement control characters. (Aw beans.)

**Comment by tigt** [comment]
Ran some of the remaining suggestions against the eBay home page:

Minification                                                       |   Bytes |            𝚫 |           𝚫%
-------------------------------------------------------------------|--------:|-------------:|-------------:
_(none)_                                                           | 388,024 |            — |            —
Omit default attributes (also shrinks JS)                          | 387,661 |   &minus;363 | &minus;0.09%
Decode HTML entities                                               | 387,600 |   &minus;424 | &minus;0.11%
Omit optional tags (also shrinks JS, DOM size via less text nodes) | 386,250 | &minus;1,774 | &minus;0.46%
